phpunit/class-wp-get-block-css-selectors-test.php

```diff
@@ -0,0 +1,319 @@
+			array( 'root' => '.wp-custom-block-class' )
+		);
+
+		$selector = wp_get_block_css_selector( $block_type );
+		$this->assertEquals( '.wp-custom-block-class', $selector );
+	}
+
+	public function test_get_root_selector_via_experimental_property() {
+		$block_type = self::register_test_block(
+			'test/block-without-selectors',
+			null,
+			array( '__experimentalSelector' => '.experimental-selector' )
+		);
+
+		$selector = wp_get_block_css_selector( $block_type );
+		$this->assertEquals( '.experimental-selector', $selector );
+	}
+
+	public function test_default_root_selector_generation() {
+		$block_type = self::register_test_block(
+			'test/without-selectors-or-supports',
+			null,
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type );
+		$this->assertEquals( '.wp-block-test-without-selectors-or-supports', $selector );
+	}
+
+	public function test_get_duotone_selector_via_selectors_api() {
+		$block_type = self::register_test_block(
+			'test/duotone-selector',
+			array(
+				'color' => array( 'duotone' => '.duotone-selector' ),
+			),
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'duotone' );
+		$this->assertEquals( '.duotone-selector', $selector );
+	}
+
+	public function test_get_duotone_selector_via_experimental_property() {
+		$block_type = self::register_test_block(
+			'test/experimental-duotone-selector',
+			null,
+			array(
+				'color' => array(
+					'__experimentalDuotone' => '.experimental-duotone',
+				),
+			)
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'duotone' );
+		$this->assertEquals( '.experimental-duotone', $selector );
+	}
+
+	public function test_no_duotone_selector_set() {
+		$block_type = self::register_test_block(
+			'test/null-duotone-selector',
+			null,
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'duotone' );
+		$this->assertEquals( '', $selector );
+	}
+
+	public function test_get_feature_selector_via_selectors_api() {
+		$block_type = self::register_test_block(
+			'test/feature-selector',
+			array(
+				'root'       => '.root',
+				'typography' => array( 'root' => '.typography' ),
+			),
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'typography' );
+		$this->assertEquals( '.typography', $selector );
+	}
+
+	public function test_get_feature_selector_via_selectors_api_shorthand_property() {
+		$block_type = self::register_test_block(
+			'test/shorthand-feature-selector',
+			array(
+				'root'       => '.root',
+				'typography' => '.typography',
+			),
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'typography' );
+		$this->assertEquals( '.typography', $selector );
+	}
+
+	public function test_no_feature_level_selector_via_selectors_api() {
+		$block_type = self::register_test_block(
+			'test/null-feature-selector',
+			array( 'root' => '.fallback-root-selector' ),
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'typography' );
```

Comment:
Trying to wrap my head about the idea of "fallback" selectors and whether consumers should have to deal with this (aka `$fallback` parameter of the function). :thinking: 

I see how different properties may have different behavior. For example, `color.background` I'll expect to fall back to `color` or `root` if there is no selector defined, though `color.duotone` should not. So, for a block such as:

```json
{
  "selectors": {
    "root": ".my-block"
  }
}
```

I'd expect this:

```php
wp_get_block_css_selector( $my_block_type, 'color.duotone' );  // Returns null.
wp_get_block_css_selector( $my_block_type, 'color.background' ); // Returns .my-block
```

Is that your expectation as well? Is there any case in which the same property (say, `color.background`) behaves differently depending on the block?

Reply:
👍 Thanks for questioning the `$fallback` param. This all gave me a few headaches when I was initially trying to iterate on my original approach in this PR that didn't include the global function and try to reuse it within the theme.json stylesheet generation.

> Is that your expectation as well? 

Yes. The expected returns in your example match my expectations and I believe are covered now by the global function.

> Is there any case in which the same property (say, color.background) behaves differently depending on the block?

I'm not aware of any situation in which `color.background` behaves differently depending on the block. 

That said, depending on the consumer's use case for this global function, they may or may not want the returned selector to fallback.

The way we generate styles in theme.json now is built around grouping style declarations under the selector it belongs to. In that context, we're only collecting selectors that will have style declarations. To reuse the global function there, we don't want the fallback selector. 

Third party devs though might wish to determine what selector any feature will ultimately end up under. So in this case, they would want to fallback to an appropriate selector. 

We could flip the default value of the `$fallback` param so the most likely use case for non-core consumers is to fallback. Effectively removing the need for most consumers to need to deal with it. 

Hopefully, this helps clarify the current situation. Let me know what you think. I'm happy to iterate on it until we're in a good state.